Source/VCGeneration/ConditionGeneration.cs

```diff
@@ -524,14 +525,16 @@ void ObjectInvariant()
         Contract.Invariant(cce.NonNullElements(vcResults));
       }
 
-      public readonly List<Counterexample> /*!>!*/ examples = new List<Counterexample>();
-      public readonly List<VCResult> /*!>!*/ vcResults = new List<VCResult>();
+      public readonly ConcurrentQueue<Counterexample> /*!>!*/
+        examples = new ConcurrentQueue<Counterexample>();
+      public readonly ConcurrentQueue<VCResult> /*!>!*/
+        vcResults = new ConcurrentQueue<VCResult>();
 
       public override void OnCounterexample(Counterexample ce, string /*?*/ reason)
```

Comment:
Mixing counterexamples from different iterations might be confusing. Can we avoid that?

I already find the current structure a bit confusing:

```c#
record VerificationResult(List<Counterexample> Errors, List<VCResult> VCResults);
record VCResult(List<Counterexample> Counterexamples, List<AssertCmd> Asserts, TimeSpan RunTime);
```

There's two lists of Counterexample in there!

I'd prefer to be more explicit:

```c#
record VerificationResult(List<SplitResult> SplitResults);
record SplitResult(List<AssertCmd> Asserts, List<IterationResult> IterationResult);
record IterationResult(List<Counterexample> Counterexamples, TimeSpan RunTime);
```

Reply:
I like this approach. I'll give it a try.

Reply:
Unfortunately, the implementation is a bit more intricately tied to that design than I expected. In particular, there's a notion of "recycled failing assertions" that I don't fully understand but that depends on counterexamples that aren't tied to our current notion of splits. Houdini also has some complex interactions with this list of counterexamples.

So, although I'd love to get rid of that duplicated list of counterexamples, I think it'll take a good bit more investigation to figure out how to do that without breaking things.

When using `/vcsStabilityIterations`, this does mean that errors (including counterexamples) will be reported more than once. However, after thinking about it a bit more, I think that's the right thing to do. Different iterations might produce different counterexamples, and those differences could be informative.

It is a bit unfortunate that all of these counterexamples are aggregated into a single list. However, it was already the case that counterexamples from different splits were aggregated into that list. So, although this PR doesn't fix that issue, I'd argue that it doesn't really make it worse, either.

Reply:
> When using /vcsStabilityIterations, this does mean that errors (including counterexamples) will be reported more than once. However, after thinking about it a bit more, I think that's the right thing to do.

When we previously looked at Excel reports to analyse Dafny verification stability, one of the key indicators we looked at was % of iterations that had errors. I think it would be good to be able to report the # of errors per iteration.

Also, showing errors including the detailed error message when using /vcsStabilityIterations doesn't seem useful, but if you do it, then wouldn't you want to remove duplicates? Showing duplicate errors seems only confusing.

How about we leave the list of counterexamples that is operated on by Houdini and RecycledFailedAssertions, but update the list that's only tied to Splits?

```csharp
record VerificationResult(List<Counterexample> Counterexamples, List<SplitResult> SplitResults); // here we leave in the existing list.
record SplitResult(List<AssertCmd> Asserts, List<IterationResult> IterationResult);
record IterationResult(List<Counterexample> Counterexamples, TimeSpan RunTime);
```

Reply:
Each `VCResult` has an outcome, so it'll still be easy to determine what % have errors. Any client that doesn't want the aggregate list of counterexamples can ignore it, and just use the shorter, nested lists. That aggregate list exists only for backward compatibility, and I hope we can eventually remove it, I just don't yet fully understand how to do so without breaking some existing functionality.

When it comes to what Boogie prints, my mental model of the feature, and the one I can imagine most easily communicating, is that it's like running Boogie multiple times, but interleaved. And with that model, seeing the same error multiple times makes sense.

I'm not 100% sure I understand your last comment. If what you're suggesting is that we package the aggregate list of counterexamples in the top-level `VerificationResult`, so it's not hanging out on its own, I think that makes sense.

Reply:
More specifically, I'm not quite sure what you mean by "update the list that's only tied to Splits".

Reply:
Oh, `VerificationResult` already includes the list of counterexamples, so there's nothing to do there, unless you had something different in mind with that last comment.

I haven't created a distinction between `VCResult`/`SplitResult` and `IterationResult`. Although doing so would make the data type slighlty more intuitive, it would make most of the clients of that data type more complex without a clear benefit. Grouping a list of `VCResult` by the `vcNum` field can achieve the same outcome, in cases where it's useful. I did just make a change to avoid duplicating identical lists of `AssertCmd` between `VCResults`, however.